### PR TITLE
[ENH, WIP] MNE_Analyze Dynamically reload plugins

### PR DESCRIPTION
**Pull request by juangpc**
Hi everybody,
This allows to, while the mne_analyze app is running, click a menu "File>Reload Plugins" and if new dlls is found in the plugin folder, the menus, views, controls will be loaded. 

- This is not WASM ready. Not sure how you could modify the plugin set during wasm execution.

While doing this I noticed a few problems with the actual application itself, not directly related but I think my changes do improve the code. 
- PluginManager stores the plugins in a specific order given by a ```m_iOrder``` variable that works as a hint on what position you want your plugin to appear. Depending on which order you set this var to be, and what other plugins decide, your plugin will appear higher or lower in the menus. 
- Some warnings have been addressed. ✊🏻 
- If a plugin has already been loaded it will be not loaded again to the PluginManager list. 
- When data is loaded to the PluginManager there is new AbstractPlugin flag:```m_bInitialized``` flag that reflects that state.
- When the plugin has had its ```menus```, ```Controls``` and ```Views``` loaded to the menus of the application's window, there will be three member flags to show that: ```m_bMenuAlreadyLoaded```, ```m_bControlAlreadyLoaded``` and  ```m_bViewAlreadyLoaded ```.
- Added `.cpp` file to AbstractPlugin.
- MainWindow was a global object (within MNEANALYZE ns). It is now owned by AnalyzeCore, and there is a global pointer that gets intitialized and can be looking at MainWindow. This is used for the globalLogger.
- AnalyzeCore would previously give access to pluginmanager to the GUI. I see no reason why this should be the case, since these are two completely different things. So now MainWindow and PluginManager both communicate with their controller (AnalyzeCore). 
- I saw a few ```onThatEvent``` type of methods/slots... I think that we should try to adopt a more imperative style. The user (a gui, in this case, for instance) uses an api of another class... so you just use it. So not react to an event, just use the class you are supposed to use. (example: ```OnMainWindowClose``` -> ```close``` as in The gui uses the function ```close``` provided by the MainWindow Object).

If you check the files, you'll find a few changes in Plugins themselves. I solve a few tiny typo/style problems in a few of them.

Still a few things to do, like: 
-  PluginManagers could generate a leak of the created menu objects if ```getMenus()``` is called more than once. Although This is not the responsibility of AbstractPlugin, at least don't think so.

### CODE REVIEW

File: applications/mne_analyze/mne_analyze/mainwindow.cpp [hunk 1]
```diff
@@ -77,57 +78,56 @@ using namespace DISPLIB;
 // DEFINE MEMBER METHODS
 //=============================================================================================================
 
-MainWindow::MainWindow(QSharedPointer<ANSHAREDLIB::PluginManager> pPluginManager,
-                       QWidget *parent)
+MainWindow::MainWindow(AnalyzeCore* pAnalyzeCore, QWidget *parent)
```
Comment (LorenzE) [inline applications/mne_analyze/mne_analyze/mainwindow.cpp:81]: Our [coding conventions](https://mne-cpp.github.io/pages/development/contr_style.html) ask for aligning function parameters underneath each other.

File: applications/mne_analyze/mne_analyze/mainwindow.cpp [hunk 2]
```diff
@@ -327,95 +438,21 @@ void MainWindow::createLogDockWindow()
...
-    for(AbstractPlugin* pPlugin : pPluginManager->getPlugins()) {
+    for(auto pPlugin : m_pAnalyzeCoreController->getLoadedPlugins()) {
         pPlugin->setObjectName(pPlugin->getName());
-        if(pPlugin) {
+        if(pPlugin && pPlugin->menuAlreadyLoaded() == false) {
             if (QMenu* pMenu = pPlugin->getMenu()) {
                 // Check if the menu already exists. If it does add the actions to the exisiting menu.
                 if(pMenu->title() == "File") {
                     for(QAction* pAction : pMenu->actions()) {
+#ifdef WASMBUILD
```
Comment (LorenzE) [inline applications/mne_analyze/mne_analyze/mainwindow.cpp:451]: Please mind indentation of #ifdef just as you did above

File: applications/mne_analyze/mne_analyze/mainwindow.cpp [hunk 3]
```diff
@@ -429,21 +466,28 @@ void MainWindow::createPluginMenus(QSharedPointer<ANSHAREDLIB::PluginManager> pP
...
+    {
+        setTabPosition(Qt::LeftDockWidgetArea,QTabWidget::West);
+        setTabPosition(Qt::RightDockWidgetArea,QTabWidget::East);
+        setDockOptions(QMainWindow::ForceTabbedDocks);
+    }
 
     //Add Plugin controls to the MainWindow
-    for(AbstractPlugin* pPlugin : pPluginManager->getPlugins()) {
-        if(QDockWidget* pControl = pPlugin->getControl()) {
+    for(AbstractPlugin* pPlugin :m_pAnalyzeCoreController->getLoadedPlugins()) {
```
Comment (LorenzE) [inline applications/mne_analyze/mne_analyze/mainwindow.cpp:487]: add one space after `:`

File: applications/mne_analyze/mne_analyze/mainwindow.h [hunk 4]
```diff
@@ -66,9 +68,9 @@ class QDockWidget;
 class QGridLayout;
 QT_END_NAMESPACE
 
-namespace ANSHAREDLIB {
-    class PluginManager;
-}
+//namespace ANSHAREDLIB {
+//    class PluginManager;
+//}
```
Comment (LorenzE) [inline applications/mne_analyze/mne_analyze/mainwindow.h:73]: can this be removed?

File: applications/mne_analyze/mne_analyze/mainwindow.h [hunk 5]
```diff
@@ -103,12 +105,17 @@ class MainWindow : public QMainWindow
     /**
      * Constructs a MainWindow which is a child of parent.
      *
+<<<<<<< HEAD
      * @param[in] pPluginManager Pointer to the plugin manager. It is needed to display subwindows froms plugins.
      * @param[in] parent Pointer to parent widget; If parent is Q_NULLPTR, the new MainWindow becomes a window.
+=======
+     * @param [in] pAnalyzeCore Pointer to the application controller class. It is needed to display subwindows froms plugins.
+     * @param [in] parent Pointer to parent widget; If parent is Q_NULLPTR, the new MainWindow becomes a window.
+>>>>>>> d7cfc0499 (rename log structure mainwindow stores analyze core not pluginmanager)
```
Comment (LorenzE) [inline applications/mne_analyze/mne_analyze/mainwindow.h:114]: Remove git conflict indicators

**Review by LorenzE — CHANGES_REQUESTED** [review]
Just some minor comments

**Review by gabrielbmotta — COMMENTED** [review]
I'm running into an issue where the reloaded plugin's controls get added underneath the others instead of joining the stack of controls
pluginload

**Comment by juangpc** [comment]
Thanks for checking this out. Interesting. I'll research about it.

File: applications/mne_analyze/libs/anShared/Management/pluginmanager.cpp [hunk 6]
```diff
@@ -70,7 +72,7 @@ PluginManager::PluginManager(QObject *parent)
 
 PluginManager::~PluginManager()
 {
-    for(AbstractPlugin* plugin : m_qVecPlugins)
+    for(AbstractPlugin*& plugin : m_qVecPlugins)
```
[the remark this answers is not in the document]
Comment (juangpc) [reply applications/mne_analyze/libs/anShared/Management/pluginmanager.cpp:75]: Agree, but I'd like to leave it as is. The addition of the reference just makes the compiler deal with references and avoids a warning.

File: applications/mne_analyze/libs/anShared/Plugins/abstractplugin.h [hunk 7]
```diff
@@ -152,27 +159,102 @@ class ANSHAREDSHARED_EXPORT AbstractPlugin : public QObject
...
     //=========================================================================================================
     /**
-     * Initializes the plugin based on cmd line inputs given by the user.
+     * Called by the EventManager in case a subscribed-for Event has happened.
      *
+<<<<<<< HEAD
      * @param[in] sArguments  the cmd line arguments.
+=======
+     * @param e The Event that has taken place
+>>>>>>> 0ecfd5a34 (define qabstractplugin methods as non virtual when there is no need for that)
```
[the remark this answers is not in the document]
Comment (juangpc) [reply applications/mne_analyze/libs/anShared/Plugins/abstractplugin.h:168]: done

**Review by juangpc — COMMENTED** [review]
I think we've addressed all the comments.

**Review by gabrielbmotta — APPROVED** [review]

File: applications/mne_analyze/libs/anShared/Management/pluginmanager.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (LorenzE) [reply applications/mne_analyze/libs/anShared/Management/pluginmanager.cpp:75]: is there also a warning for `atuo&`

**Review by LorenzE — APPROVED** [review]
sry I totally forgot about this.

**Pull request merged**